src/utils/util.ts

```diff
@@ -179,13 +191,128 @@ export const newCompilerConfig = (
 
 export const extractCompilerVersion = (
   config: HardhatConfig,
-  sourcePath?: string
+  sourcePath?: string,
+  versionPragma?: string
 ): string => {
   if (
     sourcePath !== undefined &&
     config.solidity.overrides[sourcePath] !== undefined
   ) {
     return config.solidity.overrides[sourcePath].version;
   }
+  if (versionPragma !== undefined) {
+    for (const compiler of config.solidity.compilers) {
+      if (compareVersions(compiler.version, versionPragma)) {
+        return compiler.version;
+      }
+    }
+  }
```

Comment:
And can you describe what problem you try to solve here? I need context to review this part of the code.